### DISCUSSION TITLE
MagickReadImage() memory grow

### DISCUSSION THREAD

**Discussion by 1baifeng1** [ImageMagick/ImageMagick#5158] [Development]
I observed that when MagickReadImage()/MagickReadImageBlob() are called, the memory of my process (use process-a to tag it) has been growing every times. 
Then I wrote a program to verify it. 
```#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>

#include<memory>

#include<MagickWand/MagickWand.h>

using namespace std;

class Image2D
{
private:
    /* data */
    MagickWand* image { nullptr };
public:
    Image2D(MagickWand* magick);
    ~Image2D();
};
Image2D::Image2D(MagickWand* magick)
{
    image = magick;
}

Image2D::~Image2D()
{
    if (image)
    {
        DestroyMagickWand(image);
        image = nullptr;
    }
}

int main(void)
{
    MagickWand *magick_wand;
    MagickWandGenesis();

    {
        magick_wand = NewMagickWand();
        MagickReadImage(magick_wand, "/home/fmserver/hello_magick/fruit1.png");
        sleep(3);
        DestroyMagickWand(magick_wand);
    }
    sleep(5);

    {
        Image2D* img = nullptr;
        magick_wand = NewMagickWand();
        MagickReadImage(magick_wand, "/home/fmserver/hello_magick/fruit1.png");
        img = new Image2D(magick_wand);
        MagickSetFormat(magick_wand, "PNG");
        if (img != nullptr)
        {
            sleep(3);
        }
        delete img;
    }
    sleep(5);

    {
        magick_wand = NewMagickWand();
        MagickReadImage(magick_wand, "/home/fmserver/hello_magick/fruit1.png");
        Image2D *img = new Image2D(magick_wand);
        MagickSetFormat(magick_wand, "PNG");
        sleep(3);
        delete img;
    }
    sleep(5);

    {
        magick_wand = NewMagickWand();
        MagickSetFormat(magick_wand, "PNG");
        MagickReadImage(magick_wand, "/home/fmserver/hello_magick/fruit1.png");
        MagickResetIterator(magick_wand);
        Image2D *img = new Image2D(magick_wand);
        sleep(3);
        delete img;
    }
    sleep(5);

    {
        magick_wand = NewMagickWand();
        // ClearMagickWand(magick_wand);
        MagickReadImage(magick_wand, "/home/fmserver/hello_magick/fruit1.png");
        Image2D* img = new Image2D(magick_wand);
        MagickResetIterator(magick_wand);
        MagickSetFormat(magick_wand, "PNG");
        sleep(3);
    }
    sleep(5);

    MagickWandTerminus();
    printf("finished\n");

    return 0;
}
```
I found that the first time call MagickReadImage(), the memory grow and DestroyMagickWand() can release the memory. 
But the second time, after the memory grows, DestroyMagickWand() didn't release the memory. It had same behavior as my process-a. 
From the third time, the memory of this program sometimes grows and sometimes not grow. But my process-a's memory grows every time.
I use valgrind to check if there is memory leak, but valgrind reports no memory leak.
**So I want to know, why is there such a phenomenon? And how can I keep my process-a memory from growing?**

I'm using MagickWand-7.Q16HDRI, and I also checked MagickWand-7.Q8 has same behavior.
Can anyone help me? Thanks.

**Comment by snibgo** [comment]
> I found that the first time call MagickReadImage(), the memory grow and DestroyMagickWand() can release the memory.
But the second time, after the memory grows, DestroyMagickWand() didn't release the memory.

What "second time"? In main(), there is only one "DestroyMagickWand(magick_wand);" So you read an image into memory (magick_wand), and correctly destroy it before reading again. And then you read into that wand three times without destroying it.

**Reply by 1baifeng1** [reply]
I destroy the magick_wand in `Image2D::~Image2D()` . I found that whether it is called in `main()` or in this way, the phenomenon is the same.

**Comment by urban-warrior** [comment]
Valgrind, a memory debugger, reports "definitely lost: 8 bytes in 1 blocks" suggesting the images are freed as expected.   If an image is freed, it does not necessarily mean the memory will automatically be reclaimed by the system.  Some memory allocators under specific conditions might reclaim memory, but its not guaranteed.

**Reply by 1baifeng1** [reply]
Yes, maybe it's the reason. But I think I need more knowledge about Linux memory reclaim. Thanks.
